pkg/registry/networking/servicecidr/strategy_test.go

```diff
@@ -15,3 +15,117 @@ limitations under the License.
 */
 
 package servicecidr
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/kubernetes/pkg/apis/networking"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+func TestNamespaceScoped(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	if strategy.NamespaceScoped() {
+		t.Errorf("Expected ServiceCIDRStrategy to be cluster-scoped")
+	}
+}
+
+func TestGetResetFields(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	fields := strategy.GetResetFields()
+	status := fieldpath.NewSet(
+		fieldpath.MakePathOrDie("status"),
+	)
+	if !fields["networking/v1alpha1"].Equals(status) {
+		t.Errorf("Expected 'status' field to be reset for networking/v1alpha1")
+	}
+}
+
+func TestPrepareForCreate(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	obj := &networking.ServiceCIDR{Status: networking.ServiceCIDRStatus{}}
+	strategy.PrepareForCreate(context.TODO(), obj)
+	if !reflect.DeepEqual(obj.Status, networking.ServiceCIDRStatus{}) {
+		t.Errorf("Expected status field to be cleared on create")
+	}
+}
+
+func TestPrepareForUpdate(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	newObj := &networking.ServiceCIDR{
+		Spec: networking.ServiceCIDRSpec{
+			CIDRs: []string{"10.10.0.0/24"},
+		},
+	}
+	oldObj := &networking.ServiceCIDR{}
+	strategy.PrepareForUpdate(context.TODO(), newObj, oldObj)
+	// Assert that cleared fields are now empty/unset
+}
+
+func TestValidate(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	obj := &networking.ServiceCIDR{Spec: networking.ServiceCIDRSpec{CIDRs: []string{"bad cidr"}}}
+	errors := strategy.Validate(context.TODO(), obj)
+	if len(errors) == 0 {
+		t.Errorf("Expected validation errors for invalid object")
+	}
+}
+
+func TestValidateUpdate(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	newObj := &networking.ServiceCIDR{Spec: networking.ServiceCIDRSpec{CIDRs: []string{"bad cidr"}}}
+	oldObj := &networking.ServiceCIDR{}
+	errors := strategy.ValidateUpdate(context.TODO(), newObj, oldObj)
+	if len(errors) == 0 {
+		t.Errorf("Expected validation errors for invalid update")
+	}
```

Comment:
is this not an error?

Reply:
No?

Reply:
it should be, the newObj contains an invalud cidr  🤔

Reply:
> it should be, the newObj contains an invalud cidr 🤔

The assertion fails if the length of error is zero in this test. This is as expected

Reply:
oh, lol ... then assert on the exact number of errors

Reply:
Yes

Reply:
this is still missing, I think it should be something as

```
if len(errors) != 1 { 
  t.Errorf("Expected 1 validation errors for invalid update, got %d", len(errors))
```

Reply:
Making precise assertions is not easy to achieve.
If the test code is
```go
oldObj := newServiceCIDR()
newObj := oldObj.DeepCopy()
newObj.Spec.CIDRs = []string{"bad cidr"}
if len(errors) != 1 {
	t.Errorf("Expected 1 validation errors for invalid update, got %d, errors: %v", len(errors), errors)
}
```
it got a log as
```shell
=== RUN   TestServiceCIDRStrategy
    /root/go/src/k8s.io/kubernetes/pkg/registry/networking/servicecidr/strategy_test.go:62: Expected 1 validation errors for invalid update, got 2, errors: [spec.cidrs[0]: Invalid value: "bad cidr": netip.ParsePrefix("bad cidr"): no '/' spec.cidrs: Invalid value: []string{"bad cidr"}: field is immutable]
--- FAIL: TestServiceCIDRStrategy (0.00s)
FAIL
FAIL    k8s.io/kubernetes/pkg/registry/networking/servicecidr   0.055s
```

Reply:
ok, then 2 errors are expected , you can create a table test here if you want or just use 

```
if len(errors) != 2 { 
  t.Errorf("Expected 2 validation errors for invalid update, got %d", len(errors))
```

Reply:
Done